x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -414,111 +343,22 @@ void deleteSnapshots(Map<String, List<SnapshotInfo>> snapshotsToDelete,
-            .execute(new LatchedActionListener<>(ActionListener.wrap(acknowledgedResponse -> {
-                    if (acknowledgedResponse.isAcknowledged()) {
-                        logger.debug("[{}] snapshot [{}] deleted successfully", repo, snapshot);
-                    } else {
-                        logger.warn("[{}] snapshot [{}] delete issued but the request was not acknowledged", repo, snapshot);
-                    }
+        client.admin().cluster().prepareDeleteSnapshot(repo, snapshot.getName()).execute(ActionListener.wrap(acknowledgedResponse -> {
                     slmStats.snapshotDeleted(slmPolicy);
                     listener.onResponse(acknowledgedResponse);
                 },
                 e -> {
-                    logger.warn(new ParameterizedMessage("[{}] failed to delete snapshot [{}] for retention",
-                        repo, snapshot), e);
-                    slmStats.snapshotDeleteFailure(slmPolicy);
-                    listener.onFailure(e);
-                }), latch));
-        try {
-            // Deletes cannot occur simultaneously, so wait for this
-            // deletion to complete before attempting the next one
-            latch.await();
-        } catch (InterruptedException e) {
-            logger.error(new ParameterizedMessage("[{}] deletion of snapshot [{}] interrupted",
-                repo, snapshot), e);
-            listener.onFailure(e);
-            slmStats.snapshotDeleteFailure(slmPolicy);
-        }
+                    try {
+                        logger.warn(new ParameterizedMessage("[{}] failed to delete snapshot [{}] for retention",
+                                repo, snapshot), e);
+                        slmStats.snapshotDeleteFailure(slmPolicy);
+                    } finally {
+                        listener.onFailure(e);
+                    }
+                }));
     }
 
     void updateStateWithStats(SnapshotLifecycleStats newStats) {
         clusterService.submitStateUpdateTask("update_slm_stats", new UpdateSnapshotLifecycleStatsTask(newStats));
     }
-
-    public static boolean okayToDeleteSnapshots(ClusterState state) {
-        // Cannot delete during a snapshot
-        if (state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY).entries().size() > 0) {
-            logger.trace("deletion cannot proceed as there are snapshots in progress");
-            return false;
-        }
-
-        // Cannot delete during an existing delete
-        if (state.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY).hasDeletionsInProgress()) {
-            logger.trace("deletion cannot proceed as there are snapshot deletions in progress");
-            return false;
-        }
-
-        // Cannot delete while a repository is being cleaned
-        if (state.custom(RepositoryCleanupInProgress.TYPE, RepositoryCleanupInProgress.EMPTY).hasCleanupInProgress()) {
```

Comment:
Technically repo cleanup still isn't a concurrent operation and neither is deleting a snapshot that is currently being restored.
I don't think these two cases are worth keeping this complexity around though. Both cases are kind of fringe and also the current way of checking for the cluster state not having anything conflicting in-progress isn't bullet proof to begin with (lots of possible races between checking the CS and actually starting the delete). => I think this is the reasonable simplification.